src/util.cc

```diff
@@ -423,21 +426,132 @@ const char* SpellcheckString(const char* text, ...) {
+    Fatal("Failed conversion to wide string");
+
+  return w_string;
+}
+
+int GetUtf8SizeNeeded(const WStringPiece& w_string) {
+  // Make sure that the string size can be casted to an int.
+  assert(w_string.size() <= (std::numeric_limits<int>::max)());
+
+  int size_needed = WideCharToMultiByte(CP_UTF8, 0, w_string.wstr_,
+                                        static_cast<int>(w_string.size()), NULL,
+                                        0, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed to get needed size of UTF-8 string");
+
+  return size_needed;
+}
+
+std::string WideToUtf8(const WStringPiece& w_string) {
+  if (w_string.len_ == 0)
+    return string();
+
+  // Make sure that the string size can be casted to an int.
+  assert(w_string.size() <= (std::numeric_limits<int>::max)());
+
+  int size_needed = GetUtf8SizeNeeded(w_string);
+
+  std::string u8_string(size_needed, '\0');
+  int u8_size_needed = WideCharToMultiByte(
+      CP_UTF8, 0, w_string.wstr_, static_cast<int>(w_string.size()),
+      &u8_string[0], size_needed, NULL, NULL);
+  if (u8_size_needed != size_needed)
+    Fatal("Failed conversion to UTF-8 string");
+
+  return u8_string;
+}
+
+void ConvertWideToUtf8(const WStringPiece& w_string, int size_needed,
+                       char* buff, size_t buffer_length) {
+  if (size_needed > (buffer_length - 1)) // Also account for \0
+    Fatal("Buffer is to small for UTF-8 string");
+
+  int u8_size_needed = WideCharToMultiByte(CP_UTF8, 0, w_string.wstr_,
+                                           static_cast<int>(w_string.size()),
+                                           buff, size_needed, NULL, NULL);
+
+  if (u8_size_needed != size_needed)
+    Fatal("Failed conversion to UTF-8 string");
+
+  // The conversion does not include the nulltermination.
+  buff[u8_size_needed] = '\0';
+}
+
+void WideToUtf8(const WStringPiece& w_string, char* buff,
+                size_t buffer_length) {
+  if (w_string.len_ == 0)
+    return;
+
+  int size_needed = GetUtf8SizeNeeded(w_string);
+  ConvertWideToUtf8(w_string, size_needed, buff, buffer_length);
+}
+
+char** convertCommandLine(int argc, wchar_t** wargv) {
+  size_t total_size = 0;
+
+  for (int i = 0; i < argc; i++) {
+    total_size += (GetUtf8SizeNeeded(wargv[i]) + 1);
+  }
+
+  size_t pointer_size = (argc + 1) * sizeof(char*);
+  char* commandBuffer = new char[pointer_size + total_size];
+  char** retvalue = reinterpret_cast<char**>(commandBuffer);
+  int index = 0;
+
+  for (int command_no = 0; command_no < argc; command_no++) {
+    char* buf = pointer_size + commandBuffer + index;
+
+    // Convert the commandline option.
+    ConvertWideToUtf8(
+        wargv[command_no],
+        GetUtf8SizeNeeded(wargv[command_no]),  // Size of the argument
+        buf,                                   // Buffer to write to
+        total_size - index);  // The size of the remaining buffer.
+
+    index = index + strlen(buf) + 1;  // Account for the null termination.
```

Comment:
you should not need to use strlen here.

You already know the size that was written into the buffer by the return value from the windows string conversion api.

strlen (and wstrlen) have unbounded runtime, because the string you measure *may not be* nul terminated.

Don't use them if there is any other option (which you have here).